tests/test-update-refspec-checkpoint.c

```diff
@@ -215,6 +215,340 @@ test_update_refspec_checkpoint (EosUpdaterFixture *fixture,
+                              1,
+                              &has_commit,
+                              &error);
+  g_assert_no_error (error);
+  g_assert_true (has_commit);
+
+  eos_test_client_has_commit (client,
+                              default_remote_name,
+                              2,
+                              &has_commit,
+                              &error);
+  g_assert_no_error (error);
+  g_assert_false (has_commit);
+
+  /* Now, lets say we screw up something and need to do an update on
+   * the old branch. Insert another commit, but this time without the
+   * metadata. */
+  eos_test_updater_insert_commit_steal_info (subserver->commit_graph,
+                                             1,
+                                             eos_test_updater_commit_info_new (3,
+                                                                               1,
+                                                                               default_collection_ref));
+  /* For completeness insert a new commit on the checkpoint branch */
+  eos_test_updater_insert_commit_steal_info (subserver->commit_graph,
+                                             3,
+                                             eos_test_updater_commit_info_new (4,
+                                                                               2,
+                                                                               next_collection_ref));
+  eos_test_subserver_update (subserver,
+                             &error);
+  g_assert_no_error (error);
+
+  /* Update the client client again. Even though we deployed
+   * the checkpoint, we should not have the new commit that
+   * came from the checkpoint branch. Instead we should
+   * have the newest commit on the non-checkpoint branch */
+  update_client (fixture, client);
+
+  eos_test_client_has_commit (client,
+                              default_remote_name,
+                              4,
+                              &has_commit,
+                              &error);
+  g_assert_no_error (error);
+  g_assert_false (has_commit);
+
+  eos_test_client_has_commit (client,
+                              default_remote_name,
+                              3,
+                              &has_commit,
+                              &error);
+  g_assert_no_error (error);
+  g_assert_true (has_commit);
+}
+
+/* Start with a commit, and then make a final commit on the first refspec
+ * which adds a new marker, such that when that commit is deployed, the updater
+ * will know to use a new refspec to upgrade with. However, say we screwed
+ * up and need to do a maintenance fix on the old branch. The commit from the
+ * old branch should be preferred on the next update such that the
+ * old refspec is still in use on reboot. However, later on we create
+ * another checkpoint commit on the newest commit in the old branch. That
+ * should take us to our new branch. */
```

Comment:
It would be a bit easier to follow this description if there was a little diagram of the commit graph, using the commit numbers used in the code. Something like (numbers and graph totally don’t represent what’s actually going on in this test):
```
1
|
2
/\
3 4
| |\
5 6 7
```